File: host/frontend/src/types/messages.ts
/**
 * TypeScript types mirroring the JSON schemas in shared/schemas/v1/.
 * The backend sends these over the WebSocket connection.
 */

export type MessageType =
  | 'register'
  | 'heartbeat'
  | 'event'
  | 'test_event'
  | 'state_transition'
  | 'result'
  | 'state_update'
  | 'device_list'
  | 'error';

export interface MessageEnvelope {
  version: string;
  type: MessageType;
  device_id?: string;
  payload: Record<string, unknown>;
}

export interface StateUpdatePayload {
  event: string;
  data: Record<string, unknown>;
}

export interface DeviceListPayload {
  devices: DeviceInfo[];
}

export interface DeviceInfo {
  device_id: string;
  connection_state: 'connected' | 'stale' | 'disconnected';
  firmware_version?: string;
  board_target?: string;
  last_seen_at?: string;
}

export interface ErrorPayload {
  code: string;
  message: string;
}

File: host/backend/ee_game_backend/models/messages.py
from pydantic import BaseModel, Field

from .enums import MessageType


class MessageEnvelope(BaseModel):
    """Base envelope for all WebSocket messages."""

    version: str = Field(..., pattern=r"^[0-9]+$")
    type: str
    device_id: str | None = None
    payload: dict = Field(default_factory=dict)


class RegisterPayload(BaseModel):
    firmware_version: str
    board_target: str


class RegisterMessage(BaseModel):
    version: str
    type: MessageType
    device_id: str = Field(..., min_length=1)
    payload: RegisterPayload


class HeartbeatPayload(BaseModel):
    timestamp_ms: int = Field(..., ge=0)


class HeartbeatMessage(BaseModel):
    version: str
    type: MessageType
    device_id: str = Field(..., min_length=1)
    payload: HeartbeatPayload


class ErrorPayload(BaseModel):
    code: str
    message: str


class ErrorMessage(BaseModel):
    version: str = "1"
    type: MessageType = MessageType.ERROR
    payload: ErrorPayload
